wrappers/src/shared_realm_cs.cpp

```diff
@@ -30,39 +31,131 @@
+    }
+    
+    inline CSharpBindingContext* get_or_set_managed_context(SharedRealm& realm, void* managed_realm_handle)
+    {
+        if (realm->m_binding_context == nullptr) {
+            realm->m_binding_context = std::unique_ptr<realm::BindingContext>(new CSharpBindingContext(managed_realm_handle));
+        }
+        
+        return static_cast<CSharpBindingContext*>(realm->m_binding_context.get());
+    }
 
-class CSharpBindingContext: public BindingContext {
-public:
-    CSharpBindingContext(void* managed_realm_handle) : m_managed_realm_handle(managed_realm_handle) {}
-
-    void did_change(std::vector<ObserverState> const&, std::vector<void*> const&) override
+    CSharpBindingContext::CSharpBindingContext(void* managed_realm_handle) : m_managed_realm_handle(managed_realm_handle)
     {
+        observed_rows = std::vector<ObserverState>();
+    }
+    
+    void CSharpBindingContext::did_change(std::vector<CSharpBindingContext::ObserverState> const& observed, std::vector<void*> const& invalidated)
+    {
+        for (auto const& o : observed) {
+            for (auto const& change : o.changes) {
+                if (change.kind == CSharpBindingContext::ColumnInfo::Kind::Set) {
+                    auto const& observed_object_details = static_cast<ObservedObjectDetails*>(o.info);
+                    notify_realm_object_changed(observed_object_details->managed_object_handle, get_property_index(observed_object_details->schema, change.initial_column_index));
+                }
+            }
+        }
+        
+        for (auto const& o : invalidated) {
+            remove_observed_row(o);
+        }
+        
         notify_realm_changed(m_managed_realm_handle);
     }
-
-private:
-    void* m_managed_realm_handle;
-};
-
+    
+    std::vector<CSharpBindingContext::ObserverState> CSharpBindingContext::get_observed_rows()
+    {
+        return observed_rows;
+    }
+    
+    void CSharpBindingContext::add_observed_row(const Object& object, void* managed_object_handle)
+    {
+        auto observer_state = BindingContext::ObserverState();
+        observer_state.row_ndx = object.row().get_index();
+        observer_state.table_ndx = object.row().get_table()->get_index_in_group();
+        observer_state.info = new ObservedObjectDetails(object.get_object_schema(), managed_object_handle);
+        observed_rows.push_back(observer_state);
+    }
+    
+    void CSharpBindingContext::remove_observed_row(void* managed_object_handle)
+    {
+        if (!observed_rows.empty()) {
+            for (auto it = observed_rows.begin(); it != observed_rows.end();) {
+                if (get_managed_object_handle(it->info) == managed_object_handle) {
+                    it = observed_rows.erase(it);
+                } else {
+                    ++it;
+                }
+            }
+        }
+    }
+    
+    void* CSharpBindingContext::get_managed_realm_handle()
```

Comment:
define in header instead so that the getter can be inlined in other translation units.